### PR TITLE
Splitted booking service to separate services

### PR DESCRIPTION
**Pull request by mingazhev**
BookingService was splitted to separate services:
- BookingRegistrationService - all about booking creation and the booking process
- BookingManagementService - all about changing the booking, Cancellation or Status refresh
- BookingChangesProcessor - a service to process changes - cancellation and confirmation. Here are payments, refuns etc.
- BookingResponseProcessor - all about processing responses from suppliers, synchronous or asynchronous

### CODE REVIEW

File: Api/Services/Accommodations/Bookings/BookingResponseProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                return;
+            }
+
+            if (bookingResponse.Status.ToInternalStatus() == booking.Status)
+            {
+                _logger.LogBookingResponseProcessSuccess(
+                    $"The booking response with the reference code '{bookingResponse.ReferenceCode}' has been successfully processed. No changes applied");
+                return;
+            }
+
```
Comment (kirillta) [inline Api/Services/Accommodations/Bookings/BookingResponseProcessor.cs:42]: Unnecessary space

File: Api/Services/Accommodations/Bookings/BookingRegistrationService.cs [hunk 2]
```diff
@@ -714,35 +426,66 @@ private BookingAvailabilityInfo ExtractBookingAvailabilityInfo(DataProviders dat
                 response.NumberOfNights,
                 dataProvider);
         }
-
-
-        private Result<T, ProblemDetails> WriteLogByResult<T>(Result<T, ProblemDetails> result, Action logSuccess, Action logFailure)
+
```
Comment (kirillta) [inline Api/Services/Accommodations/Bookings/BookingRegistrationService.cs:429]: Four empty lines one after another

**Review by kirillta — APPROVED** [review]

File: Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+using HappyTravel.Edo.Data;
+using HappyTravel.EdoContracts.Accommodations.Enums;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+
+namespace HappyTravel.Edo.Api.Services.Accommodations.Bookings
+{
+    public class BookingChangesProcessor
+    {
+        public BookingChangesProcessor(ISupplierOrderService supplierOrderService,
```
Comment (hasmikharutyunyan) [inline Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs:19]: Is this class registered for DI ?

Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs:19]: No, no of them are registered.
I'll add registration now

File: Api/Services/Accommodations/Bookings/BookingRegistrationService.cs (hunk 2 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingRegistrationService.cs:428]: Fixed

File: Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs (hunk 3 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs:19]: Fixed

File: Api/Services/Accommodations/Bookings/BookingResponseProcessor.cs (hunk 1 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingResponseProcessor.cs]: Fixed

**Review by hasmikharutyunyan — APPROVED** [review]

File: Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+
+            async Task<Result> SendNotifications()
+            {
+                var agent = await _context.Agents.SingleOrDefaultAsync(a => a.Id == booking.AgentId);
+                if (agent == default)
+                {
+                    _logger.LogWarning("Booking cancellation notification: could not find agent with id '{0}' for the booking '{1}'",
+                        booking.AgentId, booking.ReferenceCode);
+                    return Result.Success();
```
Comment (oleg1std) [inline Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs:58]: Empty line needed

File: Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                }
+
+                await _bookingMailingService.NotifyBookingCancelled(booking.ReferenceCode, agent.Email, $"{agent.LastName} {agent.FirstName}");
+                return Result.Success();
```
Comment (oleg1std) [inline Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs:62]: and here

File: Api/Services/Accommodations/Bookings/BookingManagementService.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+
+        public async Task<Result<Booking, ProblemDetails>> RefreshStatus(int bookingId)
+        {
+            var (_, isGetBookingFailure, booking, getBookingError) = await _bookingRecordsManager.Get(bookingId);
+            if (isGetBookingFailure)
+            {
+                _logger.LogBookingRefreshStatusFailure(
+                    $"Failed to refresh status for a booking with id {bookingId} while getting the booking. Error: {getBookingError}");
+                return ProblemDetailsBuilder.Fail<Booking>(getBookingError);
```
Comment (oleg1std) [inline Api/Services/Accommodations/Bookings/BookingManagementService.cs:76]: it's more readable with an empty line between

File: Api/Services/Accommodations/Bookings/BookingManagementService.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+            var referenceCode = booking.ReferenceCode;
+            var (_, isGetDetailsFailure, newDetails, getDetailsError) = await _dataProviderManager
+                .Get(booking.DataProvider)
+                .GetBookingDetails(referenceCode, booking.LanguageCode);
+
+            if (isGetDetailsFailure)
+            {
+                _logger.LogBookingRefreshStatusFailure($"Failed to refresh status for a booking with reference code: '{referenceCode}' " +
+                    $"while getting info from a provider. Error: {getBookingError}");
+                return Result.Failure<Booking, ProblemDetails>(getDetailsError);
```
Comment (oleg1std) [inline Api/Services/Accommodations/Bookings/BookingManagementService.cs:89]: empty line missed

File: Api/Services/Accommodations/Bookings/BookingManagementService.cs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+
+        private async Task<Result<VoidObject, ProblemDetails>> CancelBooking(Data.Booking.Booking booking, UserInfo user,
+            bool requireProviderConfirmation = true)
+        {
+            if (booking.Status == BookingStatuses.Cancelled)
+            {
+                _logger.LogBookingAlreadyCancelled(
+                    $"Skipping cancellation for a booking with reference code: '{booking.ReferenceCode}'. Already cancelled.");
+                return Result.Success<VoidObject, ProblemDetails>(VoidObject.Instance);
```
Comment (oleg1std) [inline Api/Services/Accommodations/Bookings/BookingManagementService.cs:108]: empty line missed

File: Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs (hunk 4 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs:58]: Fixed

File: Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs (hunk 5 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingChangesProcessor.cs:62]: fixed

File: Api/Services/Accommodations/Bookings/BookingManagementService.cs (hunk 6 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingManagementService.cs:76]: fixed

File: Api/Services/Accommodations/Bookings/BookingManagementService.cs (hunk 7 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingManagementService.cs:89]: fixed

File: Api/Services/Accommodations/Bookings/BookingManagementService.cs (hunk 8 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingManagementService.cs:108]: fixed

**Pull request merged**